In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

print("All libraries imported successfully! ✅")

All libraries imported successfully! ✅


In [2]:
metadata=pd.read_csv('/Users/avdhipagaria/Desktop/battery_health_project/cleaned_dataset/metadata.csv')
metadata.head()

,type,start_time,ambient_temperature,battery_id,test_id,uid,filename,Capacity,Re,Rct
0,discharge,[2010. 7. 21. 15. 0. ...,4,B0047,0,1,00001.csv,1.6743047446975208,NaN,NaN
1,impedance,[2010. 7. 21. 16. 53. ...,24,B0047,1,2,00002.csv,NaN,0.05605783343888099,0.20097016584458333
2,charge,[2010. 7. 21. 17. 25. ...,4,B0047,2,3,00003.csv,NaN,NaN,NaN
3,impedance,[2010 7 21 20 31 5],24,B0047,3,4,00004.csv,NaN,0.05319185850921101,0.16473399914864734
4,discharge,[2.0100e+03 7.0000e+00 2.1000e+01 2.1000e+01 2...,4,B0047,4,5,00005.csv,1.5243662105099023,NaN,NaN


In [3]:
print(f"Total Rows: {metadata.shape[0]}")
print(f"Total Columns: {metadata.shape[1]}")

Total Rows: 7565
Total Columns: 10


In [4]:
metadata.columns.tolist()

['type',
 'start_time',
 'ambient_temperature',
 'battery_id',
 'test_id',
 'uid',
 'filename',
 'Capacity',
 'Re',
 'Rct']

In [5]:
print('Test types',metadata['type'].value_counts())

Test types type
charge       2815
discharge    2794
impedance    1956
Name: count, dtype: int64


In [6]:
print("\n=== TEST TYPES (percentage) ===")
print(metadata['type'].value_counts(normalize=True).round(2))


=== TEST TYPES (percentage) ===
type
charge       0.37
discharge    0.37
impedance    0.26
Name: proportion, dtype: float64


In [7]:
print(metadata['battery_id'].unique())


<StringArray>
['B0047', 'B0045', 'B0048', 'B0046', 'B0043', 'B0032', 'B0039', 'B0040',
 'B0029', 'B0028', 'B0042', 'B0034', 'B0038', 'B0033', 'B0030', 'B0041',
 'B0027', 'B0044', 'B0036', 'B0025', 'B0026', 'B0031', 'B0049', 'B0050',
 'B0052', 'B0051', 'B0006', 'B0005', 'B0007', 'B0018', 'B0053', 'B0054',
 'B0056', 'B0055']
Length: 34, dtype: str


In [8]:
print(metadata.isnull().sum())


type                      0
start_time                0
ambient_temperature       0
battery_id                0
test_id                   0
uid                       0
filename                  0
Capacity               4771
Re                     5609
Rct                    5609
dtype: int64


EXAMINING THE DATASET

In [9]:
battery_df = pd.read_csv('/Users/avdhipagaria/Desktop/battery_health_project/cleaned_dataset/data/00001.csv')
battery_df.columns.tolist()

['Voltage_measured',
 'Current_measured',
 'Temperature_measured',
 'Current_load',
 'Voltage_load',
 'Time']

In [10]:
battery_df.head()

,Voltage_measured,Current_measured,Temperature_measured,Current_load,Voltage_load,Time
0,4.246711,0.000252,6.212696,0.0002,0.000,0.000
1,4.246764,-0.001411,6.234019,0.0002,4.262,9.360
2,4.039277,-0.995093,6.250255,1.0000,3.465,23.281
3,4.019506,-0.996731,6.302176,1.0000,3.451,36.406
4,4.004763,-0.992845,6.361645,1.0000,3.438,49.625


In [11]:
battery_df.shape

(490, 6)

In [12]:
print(f"Total Rows: {battery_df.shape[0]}")
print(f"Total Columns: {battery_df.shape[1]}")

Total Rows: 490
Total Columns: 6


In [13]:
print(battery_df.describe())

       Voltage_measured  Current_measured  Temperature_measured  Current_load  \
count        490.000000        490.000000            490.000000    490.000000   
mean           3.475266         -0.952767              8.272423      0.957153   
std            0.284626          0.201448              1.455269      0.202690   
min            2.470612         -1.001101              5.008084      0.000200   
25%            3.336993         -0.995898              7.313213      1.000000   
50%            3.491946         -0.995207              8.184590      1.000000   
75%            3.655987         -0.994648              9.025544      1.000000   
max            4.246764          0.000252             12.376816      1.000000   

       Voltage_load         Time  
count    490.000000   490.000000  
mean       2.817939  3205.134141  
std        0.644985  1861.395346  
min        0.000000     0.000000  
25%        2.790500  1598.417750  
50%        2.945500  3200.492500  
75%        3.105750  4809

In [14]:
print(battery_df.isnull().sum())

Voltage_measured        0
Current_measured        0
Temperature_measured    0
Current_load            0
Voltage_load            0
Time                    0
dtype: int64


In [15]:
import os

data_folder = 'cleaned_dataset/data/'

all_batteries = []

for filename in sorted(os.listdir(data_folder)):
    
    if filename.endswith('.csv'):

        filepath = os.path.join(data_folder, filename)
        df = pd.read_csv(filepath)

        df['battery_id'] = filename.replace('.csv', '')

        all_batteries.append(df)

full_df = pd.concat(all_batteries, ignore_index=True)

print(f"✅ Total Rows: {full_df.shape[0]}")
print(f"✅ Total Columns: {full_df.shape[1]}")
print(f"✅ Total Batteries: {full_df['battery_id'].nunique()}")
print(f"\nFirst 5 rows:")
print(full_df.head())

✅ Total Rows: 7376834
✅ Total Columns: 14
✅ Total Batteries: 7565

First 5 rows:
   Voltage_measured  Current_measured  Temperature_measured  Current_load  \
0          4.246711          0.000252              6.212696        0.0002   
1          4.246764         -0.001411              6.234019        0.0002   
2          4.039277         -0.995093              6.250255        1.0000   
3          4.019506         -0.996731              6.302176        1.0000   
4          4.004763         -0.992845              6.361645        1.0000   

   Voltage_load    Time battery_id Sense_current Battery_current  \
0         0.000   0.000      00001           NaN             NaN   
1         4.262   9.360      00001           NaN             NaN   
2         3.465  23.281      00001           NaN             NaN   
3         3.451  36.406      00001           NaN             NaN   
4         3.438  49.625      00001           NaN             NaN   

  Current_ratio Battery_impedance Rectified_Imp

So what we have concluded from here is that the csv file sin cleaned_dataset/data folder contains the infomation per cycle(charging,discharging,and impedenece)
thats why we have so 7565 files.
Additionally if we talk about columns in general there are only 6 columns which are common in all the csv files, but 8 extra columns are present in few files and may not be present in other, thats why if that particular column will not be present in any of the file its value will be NaN.


In [16]:
print(metadata.columns.tolist())

['type', 'start_time', 'ambient_temperature', 'battery_id', 'test_id', 'uid', 'filename', 'Capacity', 'Re', 'Rct']


In [17]:
metadata.head()

,type,start_time,ambient_temperature,battery_id,test_id,uid,filename,Capacity,Re,Rct
0,discharge,[2010. 7. 21. 15. 0. ...,4,B0047,0,1,00001.csv,1.6743047446975208,NaN,NaN
1,impedance,[2010. 7. 21. 16. 53. ...,24,B0047,1,2,00002.csv,NaN,0.05605783343888099,0.20097016584458333
2,charge,[2010. 7. 21. 17. 25. ...,4,B0047,2,3,00003.csv,NaN,NaN,NaN
3,impedance,[2010 7 21 20 31 5],24,B0047,3,4,00004.csv,NaN,0.05319185850921101,0.16473399914864734
4,discharge,[2.0100e+03 7.0000e+00 2.1000e+01 2.1000e+01 2...,4,B0047,4,5,00005.csv,1.5243662105099023,NaN,NaN


So we are basically removing the .csv extension in code below in order to join metadata and full_df because we want a common file to work on and we need a common common to perform the join.

In [18]:
metadata = pd.read_csv('cleaned_dataset/metadata.csv')
metadata['file_name']=metadata['filename'].str.replace('.csv','')
print(metadata.columns.tolist())  
print(metadata.head()) 

['type', 'start_time', 'ambient_temperature', 'battery_id', 'test_id', 'uid', 'filename', 'Capacity', 'Re', 'Rct', 'file_name']
        type                                         start_time  \
0  discharge  [2010.       7.      21.      15.       0.    ...   
1  impedance  [2010.       7.      21.      16.      53.    ...   
2     charge  [2010.       7.      21.      17.      25.    ...   
3  impedance                    [2010    7   21   20   31    5]   
4  discharge  [2.0100e+03 7.0000e+00 2.1000e+01 2.1000e+01 2...   

   ambient_temperature battery_id  test_id  uid   filename  \
0                    4      B0047        0    1  00001.csv   
1                   24      B0047        1    2  00002.csv   
2                    4      B0047        2    3  00003.csv   
3                   24      B0047        3    4  00004.csv   
4                    4      B0047        4    5  00005.csv   

             Capacity                   Re                  Rct file_name  
0  1.674304744697520

So basically the full_df is nothing but just the concatenation of tall the csv file

In [19]:
full_df.head()

,Voltage_measured,Current_measured,Temperature_measured,Current_load,Voltage_load,Time,battery_id,Sense_current,Battery_current,Current_ratio,Battery_impedance,Rectified_Impedance,Current_charge,Voltage_charge
0,4.246711,0.000252,6.212696,0.0002,0.000,0.000,00001,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,4.246764,-0.001411,6.234019,0.0002,4.262,9.360,00001,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,4.039277,-0.995093,6.250255,1.0000,3.465,23.281,00001,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,4.019506,-0.996731,6.302176,1.0000,3.451,36.406,00001,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,4.004763,-0.992845,6.361645,1.0000,3.438,49.625,00001,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [20]:
full_df.shape

(7376834, 14)

## The next step that we will be performing is of joinig the two tables(full_df and metadata.csv) on a common column:
full_df -> batter_id
metadata -> file_name
both the columns has same value

In [21]:
merged_df = full_df.merge(
    metadata[['file_name', 'battery_id', 'type', 'Capacity', 'Re', 'Rct']],
    left_on='battery_id',      # column in full_df
    right_on='file_name',      # column in metadata
    how='left'                 # keep all rows from full_df
)
print(merged_df.columns.tolist())
print(merged_df.head())
print(merged_df.shape)


['Voltage_measured', 'Current_measured', 'Temperature_measured', 'Current_load', 'Voltage_load', 'Time', 'battery_id_x', 'Sense_current', 'Battery_current', 'Current_ratio', 'Battery_impedance', 'Rectified_Impedance', 'Current_charge', 'Voltage_charge', 'file_name', 'battery_id_y', 'type', 'Capacity', 'Re', 'Rct']
   Voltage_measured  Current_measured  Temperature_measured  Current_load  \
0          4.246711          0.000252              6.212696        0.0002   
1          4.246764         -0.001411              6.234019        0.0002   
2          4.039277         -0.995093              6.250255        1.0000   
3          4.019506         -0.996731              6.302176        1.0000   
4          4.004763         -0.992845              6.361645        1.0000   

   Voltage_load    Time battery_id_x Sense_current Battery_current  \
0         0.000   0.000        00001           NaN             NaN   
1         4.262   9.360        00001           NaN             NaN   
2         3

## While executing this code I got a problem i was getting the values for voltage_charge,rectified_impedence and many other columns to be NaN, i checked where i am encountering the problem by unning through test.

In [22]:
print(f"Data type: {full_df['battery_id'].dtype}")
print(f"Sample values: {full_df['battery_id'].unique()[:5]}")


Data type: str
Sample values: <StringArray>
['00001', '00002', '00003', '00004', '00005']
Length: 5, dtype: str


In [23]:
print(f"Data type: {metadata['file_name'].dtype}")
print(f"Sample values: {metadata['file_name'].unique()[:5]}")

Data type: str
Sample values: <StringArray>
['00001', '00002', '00003', '00004', '00005']
Length: 5, dtype: str


In [24]:
test_value = full_df['battery_id'].iloc[0]
print(f"Value from full_df: '{test_value}'")
print(f"Length: {len(test_value)}")

test_value2 = metadata['file_name'].iloc[0]
print(f"\nValue from metadata: '{test_value2}'")
print(f"Length: {len(test_value2)}")

# Try to find a match manually
match = metadata[metadata['file_name'] == test_value]
print(f"Rows found: {len(match)}")
print(match)

Value from full_df: '00001'
Length: 5

Value from metadata: '00001'
Length: 5
Rows found: 1
        type                                         start_time  \
0  discharge  [2010.       7.      21.      15.       0.    ...   

   ambient_temperature battery_id  test_id  uid   filename  \
0                    4      B0047        0    1  00001.csv   

             Capacity   Re  Rct file_name  
0  1.6743047446975208  NaN  NaN     00001  


In [25]:
# Clean correct merge
merged_df = full_df.merge(
    metadata[['file_name', 'battery_id', 'type', 'Capacity']],
    left_on='battery_id',      
    right_on='file_name',      
    how='left'
)

# Rename columns to make it clean
merged_df = merged_df.rename(columns={
    'battery_id_x': 'file_id',       
    'battery_id_y': 'battery_id',    
})

# Check result
print("=== MERGED SUCCESSFULLY ===")
print(f"Total Rows: {merged_df.shape[0]}")
print(f"\nFirst 5 rows:")
print(merged_df[['file_id', 'battery_id', 'type', 
                  'Capacity', 'Voltage_measured', 
                  'Temperature_measured']].head())

=== MERGED SUCCESSFULLY ===
Total Rows: 7376834

First 5 rows:
  file_id battery_id       type            Capacity  Voltage_measured  \
0   00001      B0047  discharge  1.6743047446975208          4.246711   
1   00001      B0047  discharge  1.6743047446975208          4.246764   
2   00001      B0047  discharge  1.6743047446975208          4.039277   
3   00001      B0047  discharge  1.6743047446975208          4.019506   
4   00001      B0047  discharge  1.6743047446975208          4.004763   

   Temperature_measured  
0              6.212696  
1              6.234019  
2              6.250255  
3              6.302176  
4              6.361645  


## Step 8 — Understanding Battery Cycles (B0047 Analysis)

### What We Did:
We investigated whether 00001 was truly the first cycle 
of battery B0047 before using its Capacity as the 
original/maximum capacity for SoH calculation.

### Why We Did It:
Before calculating SoH we need to know the ORIGINAL capacity
of the battery (when it was brand new). If we pick the wrong
starting point our SoH calculation will be incorrect.

### What We Found:
- Battery B0047 has 72 discharge cycles total
- File 00001 has the HIGHEST capacity (1.6743 Ah) 
- Capacity drops every single cycle (battery aging) 
- Final capacity = 1.1567 Ah = ~69% health → RECYCLE 

### Why File IDs Jump (00001 → 00005 → 00007):
Between every discharge test there are charge and 
impedance tests. So file IDs skip numbers because 
we filtered ONLY discharge rows.

### Key Takeaway:
- 00001 IS the first cycle 
- 1.6743 IS the original capacity 
- Safe to use for SoH calculation 

### Formula We Will Use:
SoH % = (Current Capacity / Original Capacity) × 100

Example:
- Original Capacity = 1.6743 (cycle 1)
- Current Capacity  = 1.1567 (cycle 72)
- SoH = (1.1567 / 1.6743) × 100 = 69.1% (RECYCLE!) 

In [29]:
b0047 = merged_df[merged_df['battery_id'] == 'B0047']

b0047_discharge = b0047[b0047['type'] == 'discharge']

b0047_discharge_sorted = b0047_discharge.sort_values('file_id')

b0047_capacity = b0047_discharge_sorted.groupby('file_id')['Capacity'].first().reset_index()

print(" B0047 DISCHARGE CYCLES IN ORDER ")
print(b0047_capacity)

 B0047 DISCHARGE CYCLES IN ORDER 
   file_id            Capacity
0    00001  1.6743047446975208
1    00005  1.5243662105099023
2    00007  1.5080762969973425
3    00009  1.4835577960067696
4    00011  1.4671391666146525
..     ...                 ...
67   00173  1.2031062978889062
68   00175  1.1897147112197684
69   00177   1.177422568842014
70   00179   1.158443027293818
71   00181  1.1567087516841796

[72 rows x 2 columns]


## So in above code we were just filtering fot one battery just to undertand the thought process, but now instead of just one battery we are filtering all the battery at once and we are storing it in a copy so our original data stays untouched.

In [32]:
discharge_df = merged_df[merged_df['type'] == 'discharge'].copy()

discharge_df = discharge_df.sort_values(['battery_id', 'file_id'])

cycle_df = discharge_df.groupby(['battery_id', 'file_id'])['Capacity'].first().reset_index()

print("=== ONE ROW PER CYCLE ===")
print(f"Total cycles: {cycle_df.shape[0]}")
print(f"\nFirst 10 rows:")
print(cycle_df.head(10))


=== ONE ROW PER CYCLE ===
Total cycles: 2794

First 10 rows:
  battery_id file_id            Capacity
0      B0005   05122  1.8564874208181574
1      B0005   05124   1.846327249719927
2      B0005   05126  1.8353491942234077
3      B0005   05128  1.8352625275821128
4      B0005   05130  1.8346455082120419
5      B0005   05132  1.8356616600675495
6      B0005   05134  1.8351461429226603
7      B0005   05136  1.8257567905665537
8      B0005   05138  1.8247738529891333
9      B0005   05140   1.824613268496936
